[MRG+1] Added partial fit method for OvR and OvO.

kaichogami opened the pull request:
- Works only on condition if all target classes dataset chunks are passed in each iteration. In case any class in some iteration is missing, we would get incorrect results. 
- This can be overcome if we store each `estimator` in dictionary and train the relevant `estimator` in each iteration. 

I am new to scikit and machine learning in general, so please go easy on me and let me know what could be done! 
Thank you!

rvraghav93 commented:
> Works only on condition if all target classes dataset chunks are passed in each iteration.

That is how our partial fit api works. You are on the right diirection

rvraghav93 commented:
Actually you should modify it as required to be passed only at the first iteration. See how `partial_fit` is implemented in other classes.

rvraghav93 commented on sklearn/multiclass.py:
```diff
@@ -381,6 +451,7 @@ class OneVsOneClassifier(BaseEstimator, ClassifierMixin, MetaEstimatorMixin):
     def __init__(self, estimator, n_jobs=1):
         self.estimator = estimator
         self.n_jobs = n_jobs
+        self.first_partial_run = True
```
you can make this private (`_partialfit_first_run`)

rvraghav93 commented on sklearn/tests/test_multiclass.py:
```diff
@@ -426,6 +437,16 @@ def test_ovo_fit_predict():
+    ovo.partial_fit(X[100:], y[100:])
+    pred = ovo.predict(X)
+    assert_equal(len(ovo.estimators_), n_classes * (n_classes - 1) / 2)
+    assert_greater(np.mean(y == pred), 0.65)
```
You can compare this with a ful fit maybe?

kaichogami commented:
@rvraghav93 Sorry for the late reply. I modified it so that it doesn't require all classes to be passed in every iteration. Have a look and let me know your views!

rvraghav93 commented on sklearn/multiclass.py:
```diff
@@ -305,11 +355,11 @@ def decision_function(self, X):
     def multilabel_(self):
         """Whether this is a multilabel classifier"""
-        return self.label_binarizer_.y_type_.startswith('multilabel')
+        return True if self.n_classes > 2 else False
```
Why this change? `'multiclass' != 'multilabel'` correct?

kaichogami replied:
Right. They are pretty confusing. I'll change that.

rvraghav93 commented:
Thanks for addressing the comments! Pl. make the tests pass.

kaichogami commented:
@rvraghav93 should I merge all the commits as one?

rvraghav93 commented:
Yes please :)

kaichogami commented:
Alright, looks good now!

rvraghav93 commented on sklearn/multiclass.py:
```diff
@@ -74,6 +74,16 @@ def _fit_binary(estimator, X, y, classes=None):
+    """Partially fit a single binary estimator."""
+    if (not hasattr(estimator, "partial_fit")):
+            raise ValueError("Estimator must have partial_fit method")
+
```
Could you remove this line?

kaichogami replied:
Are we removing this `exception` because its superficial?

rvraghav93 replied:
no no the blank line I meant ;)

kaichogami replied:
ohh! haha! :p

rvraghav93 commented on sklearn/tests/test_multiclass.py:
```diff
@@ -426,6 +439,21 @@ def test_ovo_fit_predict():
+    pred2 = ovo2.predict(X)
+    assert_equal(len(ovo1.estimators_), n_classes * (n_classes - 1) / 2)
+    assert_greater(np.mean(y == pred1), 0.65)
+    assert_equal(np.mean(pred1 == y), np.mean(pred2 == y))
```
You can directly compare the `pred1

rvraghav93 commented on sklearn/tests/test_multiclass.py:
```diff
@@ -73,6 +75,17 @@ def test_ovr_fit_predict():
+    ovr.partial_fit(X[100:], y[100:])
+    pred = ovr.predict(X)
+    assert_equal(len(ovr.estimators_), len(np.unique(y)))
+    assert_greater(np.mean(y == pred), 0.65)
```
Here also you could add a test comparing the `pred1` and `pred2` in case of `partial_fit` / full `fit`

rvraghav93 commented:
This looks like neat work to me! Thanks! Could you address the comments alone, so I could have a final look at this?

kaichogami commented:
@rvraghav93  Done! Let me know if anything else is still missing!

rvraghav93 commented on sklearn/tests/test_multiclass.py:
```diff
@@ -426,6 +443,21 @@ def test_ovo_fit_predict():
+    ovo2.fit(X, y)
+    pred2 = ovo2.predict(X)
+    assert_equal(len(ovo1.estimators_), n_classes * (n_classes - 1) / 2)
+    assert_greater(np.mean(y == pred1), 0.65)
```
You could use the score function maybe? (No strong opinion)

kaichogami replied:
They are basically doing the same thing, i.e finding mean by comparing `real_y` and `predicted_y`, so I don't think it is necessary to do that. Infact I believe using `assert_greater(np.mean(y == pred1), 0.65)` is more explicit than `ovo1.score(X,y)`.

rvraghav93 replied:
Agreed!

rvraghav93 commented on sklearn/tests/test_multiclass.py:
```diff
@@ -73,6 +75,21 @@ def test_ovr_fit_predict():
+
+    assert_almost_equal(pred, pred2)
+    assert_equal(len(ovr.estimators_), len(np.unique(y)))
+    assert_greater(np.mean(y == pred), 0.65)
```
Same here

kaichogami committed:
Added partial_fit method for ovo and ovr
FIX issue #4167 implementing online learning where base estimators have patial_fit method.

rvraghav93 commented:
Ok. This looks fine to me. Wait for reviews from @mblondel and/or @jnothman

GaelVaroquaux commented on sklearn/multiclass.py:
```diff
@@ -74,6 +74,14 @@ def _fit_binary(estimator, X, y, classes=None):
+def _partial_fit_binary(estimator, X, y, classes=None):
+    """Partially fit a single binary estimator."""
+    if (not hasattr(estimator, "partial_fit")):
+            raise ValueError("Estimator must have partial_fit method")
```
It would be great to display something about the estimator (for instance the repr of this estimator) in the error message. It really helps users figure out what is wrong.

GaelVaroquaux commented on sklearn/multiclass.py:
```diff
@@ -203,6 +213,49 @@ def fit(self, X, y):
+        if self._first_partial_run:
+            if classes is None:
+                raise ValueError("array of all classes of target varialble"
+                                  "must be provided")
```
I think that the discoverability of the classes argument should be improved:
  * the error message above should make it clear that the classes argument is only for the first call
  * the docstring of fit should mention this argument, and its role (as eg in the docstring of the SGDClassifier).

GaelVaroquaux replied:
Also, wouldn't it be useful / desirable to reuse utils.multiclass._check_partial_fit_first_call ?

GaelVaroquaux commented on sklearn/multiclass.py:
```diff
@@ -203,6 +213,49 @@ def fit(self, X, y):
+        self.label_binarizer_ = LabelBinarizer(sparse_output=True)
+        Y = self.label_binarizer_.fit_transform(y)
+        Y = Y.tocsc()
+        columns = (col.toarray().ravel() for col in Y.T)
```
I don't really understand: why are we using a sparse_output for the LabelBinarizer and converting to arrays right after. I don't see the benefit of using a sparse output and densifying. Would it be simpler and more efficient to not use sparse_output on the binarizer and to do simple slicing to retrieve the column.

GaelVaroquaux replied:
Or is the fact that this is an iterator a trick to use less memory at a given time? I which case a comment would be required to guide the reader.

MechCoder replied:
See: https://github.com/scikit-learn/scikit-learn/issues/5620

GaelVaroquaux commented on sklearn/multiclass.py:
```diff
@@ -409,6 +476,46 @@ def fit(self, X, y):
+        """
+        if self._first_partial_run:
+            if classes is None:
+                raise ValueError("Must contain all classes of target")
```
Same remark about dostring of the function and error message. This error message is a bit hard to understand outside of the context.

GaelVaroquaux commented on sklearn/multiclass.py:
```diff
@@ -409,6 +476,46 @@ def fit(self, X, y):
+        check_consistent_length(X, y)
+        self.estimators_ = Parallel(n_jobs=self.n_jobs)(
+            delayed(_partial_fit_ovo_binary)(
+                self.estimators_.pop(0), X, y, self.classes_[i],
```
I am a bit uncomfortable with the "popping" of the estimators. Is it really required? I would rather have something like:

<pre>
...
for estimators, (class1, class2)
in itertools.izip(self.estimators, itertools.product(self.classes_, self.classes_))
</pre>

kaichogami replied:
@GaelVaroquaux "popping" is not really required. I used it to ensure correct estimator passes for correct pair of `classes`.
Also `itertools.product` creates a permutation which is not needed. Using `set(itertools.product(...))` wouldn't have worked as well so I created a generator using
`((i, j) for i in range(self.n_classes_) for j in range(i + 1, self.n_classes_))` which works as intended. I hope thats alright.

GaelVaroquaux commented on sklearn/tests/test_multiclass.py:
```diff
@@ -73,6 +75,21 @@ def test_ovr_fit_predict():
+    assert_almost_equal(pred, pred2)
+    assert_equal(len(ovr.estimators_), len(np.unique(y)))
+    assert_greater(np.mean(y == pred), 0.65)
+
```
It would be nice to have a test of when minibatches don't contain the whole set of classes, as this is the really tricky thing to code.

GaelVaroquaux commented on sklearn/tests/test_multiclass.py:
```diff
@@ -426,6 +443,21 @@ def test_ovo_fit_predict():
+    assert_equal(len(ovo1.estimators_), n_classes * (n_classes - 1) / 2)
+    assert_greater(np.mean(y == pred1), 0.65)
+    assert_almost_equal(pred1, pred2)
+
```
Same thing here:
It would be nice to have a test of when minibatches don't contain the whole set of classes, as this is the really tricky thing to code.

kaichogami committed:
Added doc, tests
Added test for partial_fit when mini-batch doesn't have all target classes. Changed doc for better explaination of the use of  paramter in partial_fit method. Used generator for passing of classes in OvO in Parallel processing instead of estimator.pop().

GaelVaroquaux commented on sklearn/multiclass.py:
```diff
@@ -203,6 +215,55 @@ def fit(self, X, y):
+        columns = (col.toarray().ravel() for col in Y.T)
+        # In cases where individual estimators are very fast to train setting
+        # n_jobs > 1 in can results in slower performance due to the overhead
+        # of spawning threads.  See joblib issue #112.
```
Issue https://github.com/joblib/joblib/issues/112 has been closed, as joblib now has specific logic to delay with this. Should this comment be removed?

GaelVaroquaux commented:
Aside from my 'comment on the comment', all looks good to me. Thanks a lot, +1 to merge (but ideally address the comment on the joblib issue first.

kaichogami committed:
Removed joblib comment, made compatible for python3
Removed the outdated comment about joblib overahead of thread spawining issue. Made izip compatible for both python2 and python3.

kaichogami commented:
@GaelVaroquaux Alright! All done now! :)

MechCoder commented:
This duplicates https://github.com/scikit-learn/scikit-learn/pull/5513?

In that case we need to credit @phdowling as well. Please make sure that no one else is working on the same thing before issuing a pull request.

MechCoder commented on sklearn/multiclass.py:
```diff
@@ -37,6 +37,11 @@
+    from itertools import izip, product
+except:
+    from itertools import product
+    izip = zip
```
from sklearn.externals.six.moves import zip

kaichogami commented:
@MechCoder My bad. I will make sure to take a closer look on pull requests next time. 
Since @phdowling has already implemented `OvR partial_fit` it will be a good idea to merge that part of commit here. Although it might require modifications as I believe it will not handle tricky cases.
Let me know what should be done.

kaichogami committed:
Changed izip import statement, importing from six moudle

MechCoder commented on sklearn/multiclass.py:
```diff
@@ -203,6 +215,53 @@ def fit(self, X, y):
+        Parameters
+        ----------
+        X : (sparse) array-like, shape = [n_samples, n_features]
+            Data.
```
sorry for being nitpicky but can you add spaces after every parameter?

MechCoder commented on sklearn/multiclass.py:
```diff
@@ -203,6 +215,53 @@ def fit(self, X, y):
+        ----------
+        X : (sparse) array-like, shape = [n_samples, n_features]
+            Data.
+        y : (sparse) array-like, shape = [n_samples] or [n_samples, n_classes]
```
To be consistent make this `(n_samples,)`, `(n_samples, n_classes)` here and everywhere.

MechCoder commented on sklearn/multiclass.py:
```diff
@@ -74,6 +76,15 @@ def _fit_binary(estimator, X, y, classes=None):
     return estimator
 
 
+def _partial_fit_binary(estimator, X, y, classes=None):
```
seems like classes is not being used here.

MechCoder commented on sklearn/multiclass.py:
```diff
@@ -74,6 +76,15 @@ def _fit_binary(estimator, X, y, classes=None):
 
+def _partial_fit_binary(estimator, X, y, classes=None):
+    """Partially fit a single binary estimator."""
+    if (not hasattr(estimator, "partial_fit")):
```
This should just be checked once and for all under the `if _check_partial_fit_first_call` block.

MechCoder commented on sklearn/multiclass.py:
```diff
@@ -203,6 +215,53 @@ def fit(self, X, y):
+                                (self.n_classes_)]
+        
+        # A sparse LabelBinarizer, with sparse_output=True, has been shown to
+        # outpreform or match a dense label binarizer in all cases and has also
```
"outpreform" is a cool word, but my concern is that it does not exist in the English dictionary AFAIK

rvraghav93 replied:
:D

MechCoder commented on sklearn/multiclass.py:
```diff
@@ -203,6 +215,53 @@ def fit(self, X, y):
+            X, next(columns) if self.classes_[i] in
+            self.label_binarizer_.classes_ else
+            np.zeros((1, len(y))), classes =
+                     ["not %s" % self.classes_[i],
```
I think you can just do away with the classes argument, unless you want to warn the user that a certain class is not present during a certain batch_fit.

MechCoder commented on sklearn/multiclass.py:
```diff
@@ -342,6 +401,19 @@ def _fit_ovo_binary(estimator, X, y, i, j):
+    y_binary[y == i] = 0
+    y_binary[y == j] = 1
+    ind = np.arange(X.shape[0])
+    return _partial_fit_binary(estimator, X[ind[cond]], y_binary,
```
just X[cond] should work no?

MechCoder commented on sklearn/multiclass.py:
```diff
@@ -342,6 +401,19 @@ def _fit_ovo_binary(estimator, X, y, i, j):
+    y = y[cond]
+    y_binary = np.empty(y.shape, np.int)
+    y_binary[y == i] = 0
+    y_binary[y == j] = 1
```
nitpick

    y_binary = np.zeros_like(y)
    y_binary[y == j] = 1

MechCoder commented on sklearn/multiclass.py:
```diff
@@ -409,6 +481,48 @@ def fit(self, X, y):
+                                (self.n_classes_-1) // 2)]
+
+        y = np.asarray(y)
+        check_consistent_length(X, y)
```
you should use `check_classification_targets` here.

MechCoder commented on sklearn/multiclass.py:
```diff
@@ -342,6 +401,19 @@ def _fit_ovo_binary(estimator, X, y, i, j):
+    y_binary[y == j] = 1
+    ind = np.arange(X.shape[0])
+    return _partial_fit_binary(estimator, X[ind[cond]], y_binary,
+                              classes=[i, j])
```
you can drop clases

MechCoder commented:
@kaichogami its all right.

Just some minor comments. +1 once comments are addressed.

kaichogami committed:
Made few tweaks based on commemnts
changed docstring of partial_fit, added check_classification_target in ovo, removed classes parameter in _partial_fit_binary.

kaichogami commented:
@MechCoder Done now.

GaelVaroquaux commented on sklearn/multiclass.py:
```diff
@@ -213,7 +274,7 @@ def predict(self, X):
         Returns
         -------
-        y : (sparse) array-like, shape = [n_samples] or [n_samples, n_classes].
+        y : (sparse) array-like, shape = [n_samples,], [n_samples, n_classes].
```
nitpick PEP8: there should be a space after the coma

GaelVaroquaux commented:
LGTM once again. Thx

GaelVaroquaux commented:
Merging, and addressing the witespace myself

MechCoder commented:
Thanks !! Updated whats_new in https://github.com/scikit-learn/scikit-learn/commit/739616a993f27553070ad3d2cc37085886823f6d
